### ISSUE TITLE
MD001 false positive: a heading not preceded by a blank line is never registered (CommonMark allows it)

### ISSUE THREAD

**Issue by smilnovi** [markdownlint/markdownlint#611]
## Description

A heading that is **not preceded by a blank line** is not registered by MD001/MD002 at all. Because
the heading is invisible to the level tracking, the *next* heading looks like a level jump and MD001
reports a false positive.

CommonMark permits such headings: *"ATX headings need not be separated from surrounding content by
blank lines, and they can interrupt paragraphs"* ([spec 0.31.2, example
77](https://spec.commonmark.org/0.31.2/#atx-headings)).

The case I hit is an HTML comment directly above a heading, which is the normal shape of files whose
managed regions are wrapped in marker comments — every such file with a subsection trips MD001.

This may be the same underlying bug as #73, which reports MD001 firing when a `###` follows a setext
(`----`) heading with no blank line between them. Both go away if a blank line is added.

### Environment

**MDL Version**

- `mdl` **0.18.1** (installed via `gem install mdl -v 0.18.1` on `ruby:3.2-slim`)
- also reproduced on **0.17.0** (Debian package `markdownlint 0.17.0-2`, Debian forky/sid)

### Expected Behavior

`## H2` should register as a level-2 heading whether or not a blank line precedes it, so a following
`### H3` is a legal one-level increment and MD001 stays quiet.

### Actual Behavior

MD001 is reported on the `### H3`. The preceding `## H2` is not merely miscounted — it is not seen:
MD002 ("first header should be a top level header"), which fires on a document whose first heading
is `##`, stays silent, indicating no heading was registered.

Note that **MD022 does see the same heading**, so the divergence appears to be in the shared
heading-extraction step rather than in MD001 itself.

## Replication Case

```sh
printf '# H1\n\n<!-- x -->\n## H2\n\ntext\n\n### H3\n' > q.md
mdl q.md
# q.md:8: MD001 Header levels should only increment by one level at a time
```

Adding a blank line after the comment, or removing the comment, is clean — so it is the absent blank
line before `## H2`, not the comment itself.

Showing that the heading is never registered:

```sh
printf 'Foo\n## baz\n'   > c.md && mdl -r MD002 c.md   # clean      -> the H2 did NOT register
printf 'Foo\n\n## baz\n' > d.md && mdl -r MD002 d.md   # MD002 fires -> the H2 registered
```

And that another rule does see it:

```sh
printf 'Foo bar\n# baz\nBar foo\n' > e.md
mdl -r MD022 e.md   # MD022 fires on line 2
mdl -r MD001 e.md   # clean
```

Full matrix on 0.18.1:

| input | MD001 |
|---|---|
| `# H1` / blank / `<!-- x -->` / `## H2` / blank / text / blank / `### H3` | **fires** |
| same, with a blank line after the comment | clean |
| same, with no comment at all | clean |
| multi-line `<!-- a` / `b -->` before `## H2` | **fires** |
| `Misc` / `----` / `### WIP` / blank / `#### Chart`  (this is #73) | **fires** |
| that same setext case with a blank line added | clean |

## Where it seems to come from

`mdl` parses via kramdown and walks the element tree for `:header` (`lib/mdl/doc.rb`,
`find_type(:header)`). Asking kramdown directly, the two bundled parsers disagree:

```ruby
require 'kramdown'
require 'kramdown-parser-gfm'

def heads(src, parser)
  out = []
  walk = ->(e) { out << [e.type, e.options[:level]] if e.type == :header; e.children.each(&walk) }
  walk.call(Kramdown::Document.new(src, input: parser).root)
  out
end

heads("Foo\n## baz\n",   'kramdown')  # => []                 <- header not parsed
heads("Foo\n## baz\n",   'GFM')       # => [[:header, 2]]
heads("Foo\n\n## baz\n", 'kramdown')  # => [[:header, 2]]
```

`Kramdown::Parser::MarkdownLint` (`lib/mdl/kramdown_parser.rb`) subclasses
`Kramdown::Parser::Kramdown`, i.e. the parser on the left — so headings that GFM/CommonMark accept
never reach the rules.

Interestingly the same class already fixes the *analogous* case for fenced code blocks, with a
comment that describes this bug exactly:

> End paragraphs when a fenced code block starts, matching GFM behavior. Without this, fenced code
> blocks without a preceding blank line are swallowed into the paragraph.

ATX headings have the same problem and are not covered.

**One thing I checked so nobody repeats it:** adding `Kramdown::Parser::GFM::ATX_HEADER_START` to
that `PARAGRAPH_END` union — the obvious symmetry with `FENCED_CODEBLOCK_START` — is **not
sufficient on its own**; both cases above still report MD001 with it applied. `Kramdown::Parser::GFM`
also substitutes the `atx_header` parser itself (`atx_header_gfm` / `atx_header_gfm_quirk`) and sets
`paragraph_end` accordingly, so the fix presumably needs the parser substitution too. I did not get
a working patch beyond that point, so I have not proposed one — the diagnosis above is the part I am
confident in, and a genuine level jump (`# H1` then `### H3`) is still correctly reported throughout.
